src/Storages/ObjectStorage/DataLakes/DeltaLake/ExpressionVisitor.cpp

```diff
@@ -0,0 +1,750 @@
+    /// as temporary names for constant columns at first.
+    /// Actual names will be assigned to them once the result expression is formed.
+    size_t literal_counter = 0;
+    /// Result expression schema.
+    const DB::NamesAndTypesList & schema;
+
+    /// Final parsing result.
+    std::shared_ptr<DB::ActionsDAG> dag;
+    /// Intermediate parsing result.
+    std::map<size_t, DB::ActionsDAG::NodeRawConstPtrs> node_lists;
+    /// First exception thrown from visitor functions.
+    std::exception_ptr visitor_exception;
+
+    DB::ContextPtr context; /// TODO: pass query context
+
+public:
+    /// `schema` is the expression schema of result expression.
+    explicit ExpressionVisitorData(const DB::NamesAndTypesList & schema_)
+        : schema(schema_)
+        , dag(std::make_shared<DB::ActionsDAG>())
+        , context(DB::Context::getGlobalContextInstance())
+    {
+    }
+
+    DB::ContextPtr getContext() const { return context; }
+
+    /// Get result of a parsed expression.
+    std::shared_ptr<DB::ActionsDAG> getResult()
+    {
+        /// In the process of parsing `node_lists` can have size > 1,
+        /// but once parsing is finished -
+        /// it must have been formed into a single list with a single element.
+        if (node_lists.size() != 1)
+        {
+            throw DB::Exception(
+                DB::ErrorCodes::LOGICAL_ERROR,
+                "Unexpected size of a result expression: {}",
+                node_lists.size());
+        }
+        if (node_lists[0].size() != 1)
+        {
+            throw DB::Exception(
+                DB::ErrorCodes::LOGICAL_ERROR,
+                "Unexpected size of a result expression at root node: {}",
+                node_lists[0].size());
+        }
+
+        const auto & nodes = node_lists[0][0]->children;
+        if (nodes.size() != schema.size())
+        {
+            throw DB::Exception(
+                DB::ErrorCodes::LOGICAL_ERROR,
+                "Unexpected size of expression list: {} (expected: {})",
+                nodes.size(), schema.size());
+        }
+
+        /// Finalize the result in outputs.
+        auto schema_it = schema.begin();
+        for (const auto & node : nodes)
+        {
+            /// During parsing we assigned temporary const_{i} names
+            /// to constant expressions,
+            /// because we do not know their names at the moment of parsing,
+            /// but once the result expression is formed -
+            /// its schema must conform with `schema` passed to constructor of ExpressionVisitorData
+            /// (only nullability of types can differ,
+            /// because when we parse non-null values, they are assigned non-nullable types).
+            /// So we substitute constant column names here.
+            if (node->type == DB::ActionsDAG::ActionType::COLUMN)
+            {
+                auto * current_node = const_cast<DB::ActionsDAG::Node *>(node);
```

Comment:
Why use `const auto & nodes` and do `const_cast` for every node? Maybe non constant references?

Reply:
Because `nodes` is `using NodeRawConstPtrs = std::vector<const Node *>` anyway.